lib/FusionInventory/Agent/Tools/Hardware.pm

```diff
@@ -540,18 +559,56 @@ sub _getMacAddress {
     # fallback on ports addresses (IF-MIB::ifPhysAddress) if unique
     my $addresses_oid = ".1.3.6.1.2.1.2.2.1.6";
     my $addresses = $snmp->walk($addresses_oid);
+
+    # filter keeping mac with a defined and associated ip otherwise takes all macs
+    my $ips = $snmp->walk('.1.3.6.1.2.1.4.20.1.2');
+    my @macs = ();
+    @macs = grep { defined } map { $addresses->{$_} } values %{$ips}
+        if (keys(%{$ips}));
+    @macs = grep { defined } values %{$addresses}
+        unless @macs;
```

Comment:
I'd rather rewrite the code this way to make the alternative more easier to catch:
<pre>
# keep only MAC address associated with an IP address, or all MAC address if there is no IP address defined
my $ips = $snmp->walk('.1.3.6.1.2.1.4.20.1.2');
my @macs = $ips ?
    grep { defined } map { $addresses->{$_} } values %{$ips}:
    grep { defined } values %{$addresses};
</pre>